sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -562,9 +562,8 @@ object ColumnPruning extends Rule[LogicalPlan] {
       d.copy(child = prunedChild(child, d.references))
 
     case p @ Project(_, s: SerializeFromObject) if p.references != s.outputSet =>
-      val prunedSerializer = s.serializer.filter { expr =>
-        p.references.contains(expr.toAttribute)
-      }
+      val usedRefs = p.references
+      val prunedSerializer = s.serializer.filter(e => usedRefs.contains(e.toAttribute))
```

Comment:
nit: why `toAttribute`? I think it is not needed (see https://github.com/apache/spark/blob/cf133e611020ed178f90358464a1b88cdd9b7889/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeSet.scala#L79)